django/urls/resolvers.py

```diff
@@ -45,6 +45,12 @@ def __init__(self, func, args, kwargs, url_name=None, app_names=None, namespaces
         self.namespaces = [x for x in namespaces if x] if namespaces else []
         self.namespace = ':'.join(self.namespaces)
 
+        if isinstance(func, functools.partial):
+            # Unwrap partial functions.
+            self.args = func.args + args
+            self.kwargs = {**func.keywords, **kwargs}
+            self.func = func = func.func
```

Comment:
This is a breaking change for view middleware (which is passed the view, and args/kwargs) - so at the very least a release note would be needed here.

It's also a change in the [documented behaviour](https://docs.djangoproject.com/en/3.1/ref/urlresolvers/#django.urls.ResolverMatch.args), which explicitly states that args/kwargs attributes of a `ResolverMatch` are those "as parsed from the URL".

Even if we updated the documented behaviour to include this caveat I think it's probably a bad idea. At the moment `request.resolver_match.args` can be used inside any view to find the arguments that were used in the actual URL. There are various reasons why knowing the args from the URL can be very useful. Once we combine the args from the URL and the partial into one, it becomes difficult to then separate them back out again.

If we want to improve the `__repr__` behaviour, I think this kind of logic should just go inside `__repr__`.

Reply:
Ok. I'm sympathetic to the concerns about backward compatibility of `args`/`kwargs` based on how they are documented, but I trump that with the [documentation of `view_name`](https://docs.djangoproject.com/en/3.1/ref/urlresolvers/#django.urls.ResolverMatch.view_name) being incorrect in the case of using a partial function without providing a `name` in the `urlpatterns`. I'm not entirely sure how we solve that one.

I see that @felixxm has closed the ticket, but I've had a look at this a bit more and still think we should improve this as the current implementation of `__repr__` implies that the object can be reconstructed. If we cannot fix this, we should change to return something of the form `<ResolverMatch ...>` instead of `ResolverMatch(...)`. Here is an example of the current output for a partial function:

```python
ResolverMatch(
    func=functools.partial, args=(), kwargs={}, url_name=partial-view,
    app_names=[], namespaces=[], route=partial/
)
```

The first thing of note is that `%s` is used for everything instead of `%r`. This works for the container-like objects whose `__str__` defers to `__repr__`, but `url_name` and `route` are strings so it breaks. We should change all `%s` to `%r` except for `func` which uses `self._func_path`. In the case of a partial function we should pass the `__repr__` of the partial function to `func`. This will give something like the following:

```python
ResolverMatch(
    func=functools.partial(<function empty_view at 0x7f3943209040>, template_name='template.html'),
    args=(), kwargs={}, url_name='partial-view', app_names=[], namespaces=[], route='partial/',
)
```

I think that this is far enough and we won't attempt to resolve the `<function ...>` to `urlpatterns_reverse.views.empty_view`. The challenge is that we'd need to handle this much like we do in the migrations serializer, and that has bits that blow up in the case of lambda functions, etc. It's not just as simple as doing `func.func.__module__ + '.' + func.func.__name__`. If we were to make this work as much as possible, however, that would be the way to fix `view_name`.

I'll push an update and re-open the ticket based on this comment.

Reply:
This seems reasonable to me